src/OrleansRuntime/Facet/GrainConstructorFacetArguementsFactory.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans.Runtime
+{
+    internal class GrainConstructorFacetArguementsFactory
+    {
+        private static readonly Type facetFactoryType = typeof(IParameterFacetFactory<>);
+        /// <summary>
+        /// Cached constructor arguement factorys by type
+        /// TODO: consider storing in grain type data and constructing at startup to avoid runtime errors. - jbragg
+        /// </summary>
+        private readonly ConcurrentDictionary<Type, ArguementsFactory> arguementsFactorys;
+        private readonly IServiceProvider services;
+
+        public GrainConstructorFacetArguementsFactory(IServiceProvider services)
+        {
+            this.services = services;
+            arguementsFactorys = new ConcurrentDictionary<Type, ArguementsFactory>();
+        }
+
+        public Type[] ArguementTypes(Type type)
+        {
+            ArguementsFactory arguementsFactory = arguementsFactorys.GetOrAdd(type, t => new ArguementsFactory(this.services, t));
+            return arguementsFactory.ArguementTypes;
+        }
+
+        public object[] CreateArguements(IGrainActivationContext grainActivationContext)
+        {
+            ArguementsFactory arguementsFactory = arguementsFactorys.GetOrAdd(grainActivationContext.GrainType, type => new ArguementsFactory(this.services, type));
+            return arguementsFactory.CreateArguements(grainActivationContext);
+        }
+
+        /// <summary>
+        /// Facet Arguement factory
+        /// </summary>
+        private class ArguementsFactory
+        {
+            private readonly List<Factory<IGrainActivationContext, object>> arguementFactorys;
+
+            public ArguementsFactory(IServiceProvider services, Type type)
+            {
+                this.arguementFactorys = new List<Factory<IGrainActivationContext, object>>();
+                List<Type> types = new List<Type>();
+                IEnumerable<ParameterInfo> parameters = type.GetConstructors()
+                                                            .FirstOrDefault()?
+                                                            .GetParameters() ?? Enumerable.Empty<ParameterInfo>();
+                foreach (ParameterInfo parameter in parameters)
+                {
+                    var attribute = parameter.GetCustomAttribute<FacetAttribute>();
+                    if (attribute == null) continue;
+                    IFactoryBridge bridge = Activator.CreateInstance(typeof(FactoryBridge<>).MakeGenericType(attribute.GetType())) as IFactoryBridge;
+                    if (bridge == null) continue;
+                    bridge.Factory = services.GetRequiredService(facetFactoryType.MakeGenericType(attribute.GetType()));
```

Comment:
My understanding was:

- Attributes were to be treated as part of the abstraction layer, they should contain no logic.
- Feature factories were to be obtained from the container
- Attribute processing logic should use the same feature factories (obtained from the container) that the grains use.

All of these conditions were met.

I also argued:

- There is not likely a single common pattern to wire attributes for all features up to their factories.
- If we did develop a single common set of logic that would wire up attributes for all features to their factories it would likely be complicated and hard to maintain.
- If, after introducing a couple of features (3-6 months from now) and have more experience adding features, we conclude that there is a single (or a small set) of patterns that fit all of our needs, we can make them part of the framework and deprecate the IParameterFacetFactory (now called IAttributeToFactoryMapper), which is low risk.

The cost of this pattern is that for each new feature (new attribute really), a relatively small set of logic is needed (implementation of IAttributeToFactoryMapper) to wire up the feature factories (from data in the attribute) to the general factory pattern cached by the argument factory.  This, imo, is a relatively small price to pay to keep us from painting ourselves into a corner should we misjudge the conformity of features.